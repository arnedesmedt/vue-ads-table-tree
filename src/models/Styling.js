export default class Styling {
    constructor (properties = {}) {
        if (!(properties instanceof Object)) {
            throw new Error(
                'Styling first argument has to be an object with the properties rowsEven, rowsOdd, rowsAll, rowsExceptLast, columnsEven, columnsOdd, columnsAll, columnsExceptLast and/or table.' +
                ' \'' + typeof properties + '\' given.'
            );
        }

        this.rowsEven = properties.hasOwnProperty('rowsEven') ? properties.rowsEven : 'vue-ads-table-tree-bg-grey-lightest';
        this.rowsOdd = properties.hasOwnProperty('rowsOdd') ? properties.rowsOdd : 'vue-ads-table-tree-bg-white';
        this.rowsAll = properties.hasOwnProperty('rowsAll') ? properties.rowsAll : 'vue-ads-table-tree-hover:bg-grey-lighter';
        this.rowsAllExceptLast = properties.hasOwnProperty('rowsAllExceptLast') ? properties.rowsAllExceptLast : 'vue-ads-table-tree-border-b';
        this.columnsEven = properties.hasOwnProperty('columnsEven') ? properties.columnsEven : '';
        this.columnsOdd = properties.hasOwnProperty('columnsOdd') ? properties.columnsOdd : '';
        this.columnsAll = properties.hasOwnProperty('columnsAll') ? properties.columnsAll : '';
        this.columnsAllExceptLast = properties.hasOwnProperty('columnsAllExceptLast') ? properties.columnsAllExceptLast : 'vue-ads-table-tree-border-r';
        this.table = properties.hasOwnProperty('table') ? properties.table : 'vue-ads-table-tree-border';
    }

    set rowsEven (rowsEven) {
        this._rowsEven = rowsEven;
    }

    get rowsEven () {
        return this._rowsEven;
    }

    set rowsOdd (rowsOdd) {
        this._rowsOdd = rowsOdd;
    }

    get rowsOdd () {
        return this._rowsOdd;
    }

    set rowsAll (rowsAll) {
        this._rowsAll = rowsAll;
    }

    get rowsAll () {
        return this._rowsAll;
    }

    set rowsAllExceptLast (rowsAllExceptLast) {
        this._rowsAllExceptLast = rowsAllExceptLast;
    }

    get rowsAllExceptLast () {
        return this._rowsAllExceptLast;
    }

    set columnsEven (columnsEven) {
        this._columnsEven = columnsEven;
    }

    get columnsEven () {
        return this._columnsEven;
    }

    set columnsOdd (columnsOdd) {
        this._columnsOdd = columnsOdd;
    }

    get columnsOdd () {
        return this._columnsOdd;
    }

    set columnsAll (columnsAll) {
        this._columnsAll = columnsAll;
    }

    get columnsAll () {
        return this._columnsAll;
    }

    set columnsAllExceptLast (columnsAllExceptLast) {
        this._columnsAllExceptLast = columnsAllExceptLast;
    }

    get columnsAllExceptLast () {
        return this._columnsAllExceptLast;
    }

    set table (table) {
        this._table = table;
    }

    get table () {
        return this._table;
    }

    rowClasses (index, last = false) {
        let rowClasses = {};

        if (index % 2 === 0 && this.rowsEven) {
            rowClasses[this.rowsEven] = true;
        }

        if (index % 2 === 1 && this.rowsOdd) {
            rowClasses[this.rowsOdd] = true;
        }

        if (!last && this.rowsAllExceptLast) {
            rowClasses[this.rowsAllExceptLast] = true;
        }

        if (this.rowsAll) {
            rowClasses[this.rowsAll] = true;
        }

        return rowClasses;
    }

    columnClasses (index, last = false) {
        let columnClasses = {};

        if (index % 2 === 0 && this.columnsEven) {
            columnClasses[this.columnsEven] = true;
        }

        if (index % 2 === 1 && this.columnsOdd) {
            columnClasses[this.columnsOdd] = true;
        }

        if (!last && this.columnsAllExceptLast) {
            columnClasses[this.columnsAllExceptLast] = true;
        }

        if (this.columnsAll) {
            columnClasses[this.columnsAll] = true;
        }

        return columnClasses;
    }

    tableClasses () {
        let tableClasses = {};

        if (this.table) {
            tableClasses[this.table] = true;
        }

        return tableClasses;
    }
}
