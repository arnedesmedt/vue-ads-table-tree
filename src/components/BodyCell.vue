<script>
import Row from '../models/Row';
import Column from '../models/Column';
import ToggleChildrenButton from './ToggleChildrenButton';
import Styling from '../models/Styling';

export default {
    name: 'BodyCell',

    components: {
        ToggleChildrenButton,
    },

    props: {
        styling: {
            type: Styling,
            required: true,
        },

        row: {
            type: Row,
            required: true,
        },

        column: {
            type: Column,
            required: true,
        },

        columnSlot: {
            required: false,
        },

        index: {
            type: Number,
            required: true,
        },

        rowIndex: {
            type: Number,
            required: true,
        },
    },

    render (createElement) {
        return createElement('td', {
            class: this.cellClasses,
            style: this.cellStyle,
        }, [
            createElement('span', {}, this.value(createElement)),
        ]);
    },

    computed: {
        cellClasses () {
            return Object.assign(
                this.styling.columnClasses(0, this.column.last),
                {
                    'vue-ads-table-tree-px-4': true,
                    'vue-ads-table-tree-py-2': true,
                    'vue-ads-table-tree-text-sm': true,
                    ['vue-ads-table-tree-w-' + this.column.width]: true,
                }
            );
        },

        cellStyle () {
            let paddingLeft = 1 + this.first * this.row.countParents() * 1.5;

            return {
                'padding-left': paddingLeft + 'rem',
            };
        },

        first () {
            return this.index === 0;
        },
    },

    methods: {
        value (createElement) {
            let elements = [];

            if (this.first && this.row.hasChildren) {
                elements.push(
                    createElement(ToggleChildrenButton, {
                        props: {
                            expanded: this.row.showChildren,
                            loading: this.row.childrenLoading,
                        },
                        nativeOn: {
                            'click': this.toggleChildren,
                        },
                    }),
                );
            }

            if (this.columnSlot) {
                elements.push(this.columnSlot({row: this.row, index: this.rowIndex}));
            } else if (this.column.property && this.row.hasOwnProperty(this.column.property)) {
                elements.push(this.row[this.column.property]);
            }

            return elements.length > 0 ? elements : [''];
        },

        toggleChildren () {
            this.$emit('toggleChildren');
        },
    },
};
</script>
