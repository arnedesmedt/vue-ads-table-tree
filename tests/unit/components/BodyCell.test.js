import { shallowMount } from '@vue/test-utils';

import BodyCell from '../../../src/components/BodyCell';
import Row from '../../../src/models/Row';
import Column from '../../../src/models/Column';
import Styling from '../../../src/models/Styling';

describe('BodyCell', () => {
    let bodyCell;

    beforeEach(() => {
        bodyCell = shallowMount(BodyCell, {
            propsData: {
                styling: new Styling({
                    columnsAllExceptLast: 'vue-ads-table-tree-border-r',
                }),
                row: new Row({
                    firstName: 'arne',
                }),
                column: new Column({
                    property: 'firstName',
                }),
                index: 0,
            },
        });
    });

    it('returns the default cell classes', () => {
        expect(bodyCell.vm.cellClasses).toEqual({
            'vue-ads-table-tree-border-r': true,
            'vue-ads-table-tree-px-4': true,
            'vue-ads-table-tree-py-2': true,
            'vue-ads-table-tree-text-sm': true,
            'vue-ads-table-tree-w-auto': true,
        });

        expect(bodyCell.vm.first).toBeTruthy();
        expect(bodyCell.vm.cellStyle['padding-left']).toBe('1rem');

        expect(bodyCell.html()).toMatchSnapshot();
    });

    it('changes the padding if the number of parents changes', () => {
        bodyCell.setProps({
            row: new Row({
                parent: new Row(),
            }),
        });

        expect(bodyCell.vm.cellStyle['padding-left']).toBe('2.5rem');
    });

    it('adds a toggle children button if the row has children', () => {
        bodyCell.setProps({
            row: new Row({
                hasChildren: true,
            }),
        });

        expect(bodyCell.html()).toMatchSnapshot();
    });

    it('is empty with no matching properties', () => {
        bodyCell.setProps({
            row: new Row({
                lastName: 'de smedt',
            }),
        });
    });

    it('emits toggle children, when calling toggle children', () => {
        bodyCell.vm.toggleChildren();

        expect(bodyCell.emitted().toggleChildren).toBeTruthy();
    });
});
