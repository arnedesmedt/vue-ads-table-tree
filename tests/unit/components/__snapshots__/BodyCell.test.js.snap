// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyCell adds a toggle children button if the row has children 1`] = `<td class="vue-ads-table-tree-border-r vue-ads-table-tree-px-4 vue-ads-table-tree-py-2 vue-ads-table-tree-text-sm vue-ads-table-tree-w-auto" style="padding-left: 1rem;"><span><i class="fa vue-ads-table-tree-pr-2 vue-ads-table-tree-cursor-pointer fa-plus-square"></i></span></td>`;

exports[`BodyCell returns the default cell classes 1`] = `<td class="vue-ads-table-tree-border-r vue-ads-table-tree-px-4 vue-ads-table-tree-py-2 vue-ads-table-tree-text-sm vue-ads-table-tree-w-auto" style="padding-left: 1rem;"><span>arne</span></td>`;
